libkbfs/unflushed_path_cache.go

```diff
@@ -0,0 +1,328 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/keybase/client/go/logger"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/kbfs/kbfscodec"
+	"golang.org/x/net/context"
+)
+
+type unflushedPathCacheState int
+
+const (
+	upcUninitialized unflushedPathCacheState = iota
+	upcInitializing
+	upcInitialized
+)
+
+type unflushedPathsPerRevMap map[string]bool
+type unflushedPathsMap map[MetadataRevision]unflushedPathsPerRevMap
+
+// unflushedPathCache tracks the paths that have been modified by MD
+// updates that haven't yet been flushed from the journal.
+type unflushedPathCache struct {
+	lock            sync.RWMutex
+	state           unflushedPathCacheState
+	unflushedPaths  unflushedPathsMap
+	ready           chan struct{}
+	chainsPopulator chainsPathPopulator
+	appendQueue     []ImmutableRootMetadata
+	removeQueue     []MetadataRevision
+}
+
+var errUPCNotInitialized = errors.New("The unflushed path cache is not yet initialized")
+
+// getUnflushedPaths returns a copy of the unflushed path cache if it
+// has been initialized, otherwise nil.  It must be called under the
+// same lock as the callers of appendToCache/removeFromCache.
+func (upc *unflushedPathCache) getUnflushedPaths() unflushedPathsMap {
+	upc.lock.RLock()
+	defer upc.lock.RUnlock()
+	if upc.unflushedPaths == nil {
+		return nil
+	}
+	cache := make(unflushedPathsMap)
+	// Only need to deep-copy the outer level of the map; the inner
+	// level (per-revision) is immutable.
```

Comment:
Only if the callers don't modify it, right? May want to clarify / mention that in the function-level comment.